engine/src/lib.rs

```diff
@@ -209,11 +209,23 @@ mod contract {
     #[no_mangle]
     pub extern "C" fn call() {
         let io = Runtime;
-        let args: FunctionCallArgs = io.read_input_borsh().sdk_unwrap();
-        let mut engine = Engine::new(predecessor_address(), io).sdk_unwrap();
-        Engine::call_with_args(&mut engine, args)
-            .map(|res| res.try_to_vec().sdk_expect("ERR_SERIALIZE"))
-            .sdk_process();
+        let args: Option<CallArgsType> = {
+            if let Ok(value) = io.read_input_borsh::<FunctionCallArgs>() {
+                Some(CallArgsType::New(value))
+            } else if let Ok(value) = io.read_input_borsh::<FunctionCallArgsLegacy>() {
+                Some(CallArgsType::Legacy(value))
+            } else {
+                None
+            }
+        };
```

Comment:
A few comments here:

- I think the initial type to attempt to parse from the input should be `CallArgsType`. The reason is because an enum has built-in backwards compatibility for future changes we might make (the byte indicating the enum variant effectively functions as a version byte).
- It is a little more efficient to read the input into a vector (`io.read_input().to_vec()`) instead of invoking `io.read_input_borsh()` multiple times because `read_input_borsh` will read/write from the host register every time and allocate a new vector every time.
- I think this fallback logic should be attached to the `CallArgsType` itself as a function like `pub fn deserialize(input: &[u8]) -> Self`. This way we can re-use the logic in another place easily (if need).
- Nit: I don't really like the `CallArgsType` name. Having the `Type` suffix seems redundant (obviously its a type). Personally I would have it take on the `FunctionCallArgs` name, and the new struct could be called `FunctionCallArgsV1`, or something like that.

Reply:
Addressed notes and remarks here, made a commit with corrections and improvements:
https://github.com/aurora-is-near/aurora-engine/pull/351#issuecomment-969075880

Reply:
Thanks for the changes. Please also make changes for the first bullet (deserialize enum directly), and third bullet (encapsulate fallback logic in the `impl CallArgs`).

Reply:
Here I cannot agree, 'cause took ownership on this task.
Would explain my opinion:
I was wrong about enum - it's flexible for usage with `Engine::call_with_args()` for matching of already successfully parsed type, but in case with borsh we would have a type incompatibility issue, I mean we cannot do as in example you've provide [here](https://github.com/birchmd/aurora-engine/commit/92313f73dcaa0c1f5a53d8d16ebbacf89ad66849), the better way is to returned parsed bytes would be mapped on a struct, and enum would be used in type matches in further, thus I just decide and prefer to make things more simple, clean, straightforward and obvious, and readable (what is more important!), `if let` has more good fit for this usage scenario, rather than `match`, we just trying to parse to newer type, then, if borsh returned error, we fallback to try to parse and map bytes to older type, and even if in this case we would have an error (if something unusual would be sent through register), then we fallback to exception panic with message.
CallArgs enum would be useful in `Engine::call_with_args()` to choose between types, when we send parsed bytes mapped already to new OR legacy data structure already contained in the enum as collection of CallArg types.
Let's just make things more simple and clean, straightforward and obvious, and readable as well, and then we would figure out how we would reuse that pattern for future operations with borsh, sdk and registers, if we would have such case in the future, and if we would have ones, we can make better decision on refactor and generalization of that pattern for reuse.

Reply:
Sure, you can keep using `if let` if you want instead of `match`, I don't care about that much either way.

What I do care about is 

1. The first type we attempt to deserialize from the bytes is `CallArgs`. This is better than trying to deserialize a particular variant because it means no logic changes need to happen in the case we add new variants to `CallArgs` in the future. And this is something we might do, for example if we want to allow access lists, or some other new Ethereum transaction type that does not yet exist.
2. The logic for deserializing the bytes is in an impl for `CallArgs`. By encapsulating in this way it means the logic can easily be re-used in other parts of the code, and it makes the `call` function in `lib.rs` cleaner because the deserialization logic is hidden behind a function call.

Reply:
Ok, this make sense, I agree, I will reimplement this as you proposed, to address your remarks, and to feel how borsh would map bytes to structs encapsulated in enums.